File: backend/app/services/parity.py
"""Чётность недели (числитель/знаменатель).

Правило (проверено на реальных заголовках замен, см. docs/PARSER_ANALYSIS.md §2.4):
неделя, в которую попадает 1 сентября учебного года, — нечётная (числитель).
Дополнительно сервис калибруется по заголовку файла замен: если заголовок
говорит «четный», а правило даёт «нечетная» — вводится постоянный сдвиг.
"""
from __future__ import annotations

from datetime import date, timedelta

ODD = "odd"
EVEN = "even"


def base_parity(d: date) -> str:
    """Чётность по якорю «неделя 1 сентября = нечётная»."""
    sep1_year = d.year if d.month >= 8 else d.year - 1
    sep1 = date(sep1_year, 9, 1)
    anchor_monday = sep1 - timedelta(days=sep1.weekday())
    weeks = (d - anchor_monday).days // 7
    return ODD if weeks % 2 == 0 else EVEN


class ParityResolver:
    """Чётность с автокалибровкой по заголовкам файла замен."""

    def __init__(self) -> None:
        self._flip = False
        self._calibrated_on: date | None = None
        self.warnings: list[str] = []

    def calibrate(self, declared_date: date, declared_parity: str | None) -> None:
        if declared_parity not in (ODD, EVEN):
            return
        actual = base_parity(declared_date)
        flip = actual != declared_parity
        if self._calibrated_on is not None and flip != self._flip:
            msg = (f"Конфликт чётности недели: {declared_date} объявлена "
                   f"«{declared_parity}», предыдущая калибровка противоречит")
            self.warnings.append(msg)
        self._flip = flip
        self._calibrated_on = declared_date

    def parity(self, d: date) -> str:
        p = base_parity(d)
        return (EVEN if p == ODD else ODD) if self._flip else p

File: frontend/src/services/parity.ts
// Чётность недели (числитель/знаменатель).
// Правило: неделя 1 сентября = нечётная (числитель).

import type { ReplacementBlockJSON } from '../parser/types'

export const ODD = 'odd' as const
export const EVEN = 'even' as const

export function baseParity(d: Date): string {
  const year = d.getMonth() >= 7 ? d.getFullYear() : d.getFullYear() - 1
  const sep1 = new Date(year, 8, 1) // 1 сентября
  const anchorMonday = new Date(sep1)
  anchorMonday.setDate(sep1.getDate() - sep1.getDay())
  const weeks = Math.floor((d.getTime() - anchorMonday.getTime()) / (7 * 86400000))
  return weeks % 2 === 0 ? ODD : EVEN
}

export class ParityResolver {
  private flip = false
  private calibratedOn: Date | null = null
  warnings: string[] = []

  calibrate(d: Date, declaredParity: string | null) {
    if (declaredParity !== ODD && declaredParity !== EVEN) return
    const actual = baseParity(d)
    const flip = actual !== declaredParity
    if (this.calibratedOn !== null && flip !== this.flip) {
      this.warnings.push(
        `Конфликт чётности: ${d.toISOString().slice(0, 10)} объявлена «${declaredParity}»`,
      )
    }
    this.flip = flip
    this.calibratedOn = d
  }

  parity(d: Date): string {
    const p = baseParity(d)
    return this.flip ? (p === ODD ? EVEN : ODD) : p
  }
}

export function createResolverFromBlocks(blocks: ReplacementBlockJSON[]): ParityResolver {
  const r = new ParityResolver()
  for (const b of blocks) {
    r.calibrate(new Date(b.date + 'T12:00:00'), b.parity)
  }
  return r
}
